### PR TITLE
style injection on per page basis

### PR DESCRIPTION
**Pull request by luca-peruzzo**
in login.routes we can define if we want to use default css on per page basis using: `data: { doUseDefaultCss: true },`
this value is saved on `ActivatedRouteSnapshot` and can be read in `styleSheetResolver` and `KcClassPipe`

### CODE REVIEW

**Comment by luca-peruzzo** [comment]
closes #10

File: angular.json [hunk 1]
```diff
@@ -74,14 +74,7 @@
                             "buildOptimizer": false,
                             "optimization": false,
                             "sourceMap": true,
-                            "namedChunks": true,
-                            "styles": [
```
Comment (kathari00) [inline angular.json:78]: Why remove the config for local development? Have you tested the solution?

File: src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts [hunk 2]
```diff
@@ -1,13 +1,49 @@
...
+    loadStyle() {
+        let stylesheets = [
+            `${this.kcContext.url.resourcesCommonPath}/node_modules/@patternfly/patternfly/patternfly.min.css`,
+            `${this.kcContext.url.resourcesCommonPath}/node_modules/patternfly/dist/css/patternfly.min.css`,
+            `${this.kcContext.url.resourcesCommonPath}/node_modules/patternfly/dist/css/patternfly-additions.min.css`,
+            `${this.kcContext.url.resourcesCommonPath}/lib/pficon/pficon.css`,
+            `${this.kcContext.url.resourcesPath}/css/login.css`
+        ];
+        if (isDevMode()) {
+            stylesheets = stylesheets.map(s => `/static/media${s}`);
```
Comment (kathari00) [inline src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: Get the resources from assets in local mode

**Comment by kathari00** [comment]
Why make it so complex? You could also just give a doUseDefaultCss to the parent and it loads conditionally then.

File: src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts (hunk 2 above)
Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: the rule in angular.json makes a path rewrite from assets/ to static/media
```
 {
    "glob": "**/*",
    "input": "./src/keycloak-theme/assets",
    "output": "./static/media"
}
```

**Comment by luca-peruzzo** [comment]
> Why make it so complex? You could also just give a doUseDefaultCss to the parent and it loads conditionally then.

but it has to be a "per page decision". 
Maybe I want custom style on "login" but not on "error".
This is the behaviour @garronej has implemented

**Comment by kathari00** [comment]
Yes that is exactly what you are doing when just giving this information to the parent. 

It is crucial to know that this is not a full SPA anymore. The login form is a plain html post form which does a server redirect. This is something that should not be done with an SPA but it is necessary (that is why I made the redirect-handler for Angular Material). So, everytime we change the page by a server redirect (always) we re-load everything.

**Comment by luca-peruzzo** [comment]
> Yes that is exactly what you are doing when just giving this information to the parent.
> 
> It is crucial to know that this is not a full SPA anymore. The login form is a plain html post form which does a server redirect. This is something that should not be done with an SPA but it is necessary (that is why I made the redirect-handler for Angular Material). So, everytime we change the page by a server redirect (always) we re-load everything.

login page has only one parent and it is the main router outlet. It is the same parent of error, register etc...
So if I want to use a custom css only on the login page, how can I have a different configuration?
I need a resolver to run before the route is injected into the router outlet to read this value and apply or not apply the default css.

File: src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts (hunk 2 above)
Comment (kathari00) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: But this is only relevant for Production mode? Have you tried this?

Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: it isn't.
it is defined in build.options.
you can override configurations in configurations.production, configurations.development etc...

Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: if you want to try just run npm serve with these changes (I've already do it on local and on docker)

Comment (kathari00) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: no it does not work in locale

Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: image

Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: it works... have you run `npx keycloakify copy-keycloak-resources-to-public`?

Comment (luca-peruzzo) [reply src/keycloak-theme/login/common/services/dynamic-style-loader.service.ts:27]: if you have errors on compilation or at runtime please paste them here

**Comment by kathari00** [comment]
I have invited you as a collaborator to this repo. Are you on Discord?

**Comment by luca-peruzzo** [comment]
> I have invited you as a collaborator to this repo. Are you on Discord?

yes! I'm luca_p_

**Pull request merged**